### ISSUE TITLE
Signatures aggregation service

### ISSUE THREAD

**Issue by rimrakhimov** [blockscout/blockscout-rs#133]
Currently there are several solutions that implement Ethereum signature databases and provides an API to search for the functions and events by their selectors (https://sig.eth.samczsun.com/, https://www.4byte.directory/). Those databases may contain very similar but not identical information. We would like to implement a service that a) makes a call to those databases to search for corresponding methods, aggregates responses and return aggregated data to the caller; b) submits new ABIs to those databases.

**Comment by rimrakhimov** [comment]
Endpoint (a) could be called from the main blockscout instance when it needs to decode a transaction of unverified contract.

**Comment by rimrakhimov** [comment]
Regarding endpoint (b) we have considered 3 possible options:

1) The aggregation service is responsible only for importing ABIs into Ethereum Signature Databases. Aggregation service is called by Blockscout instance after smart-contract-verifier returns verification result.
![smart-contract-verifier-proxy-1](https://user-images.githubusercontent.com/25086020/193656665-e8434488-738b-4531-a623-58e0c74ab329.svg)

**Comment by rimrakhimov** [comment]
## Endpoint (b)

### Option 2

The _aggregation service_ is responsible only for importing ABIs into Ethereum Signature Databases. But data into `signatures-aggregator` is submitted by `smart-contract-verifier` directly. The `Blockscout` instance does not know anything about `signatures-aggregator` and requires no changes at all. 

![smart-contract-verifier-proxy-2](https://user-images.githubusercontent.com/25086020/193679208-6f873d70-49b0-497c-9fa5-8c510b0669cc.svg)

**Comment by rimrakhimov** [comment]
## Endpoint (b)

### Option 3

In the third option, we have another microservice named `smart-contract-verifier-proxy`. In comparison with the second option, that service allows `smart-contract-verifier` not to to be changed and allows it to be  _"pure"_ in the sense that it just accepts requests from the clients and returns responses. That approach have one interesting 
![smart-contract-verifier-proxy-3](https://user-images.githubusercontent.com/25086020/193681473-73fcb594-b3bb-41ce-ac82-f36329424c82.svg)

**Comment by sevenzing** [comment]
I suggest considering a few other short names for this service:

+ eth-sig-proxy
+ sig-documenter
+ sig-storage
+ sig-provider

**Comment by rimrakhimov** [comment]
@sevenzing I personally like _sig-provider_ the most, but I suggest to create a poll inside a Blockscout team regarding the naming, when the service functionality would be finalized

**Comment by vbaranov** [comment]
In regards, Endpoint (b) Option 2

> The approach has a disadvantage, though, that smart-contract-verifier service lose its "puriness" - the fact that its all functionality is to accept requests from the clients, compile and verify contracts, and return responses back without any additional interactions.

We could make that integration with sig databases optional, and the base `smart-contract-verifier` service will still be "pure".

Anyways, Endpoint (b) Option 2/Option 3 look more preferable than Option 1 since Option 1 requires efforts from 2 teams (Elixir/Rust). And both Option 2/Option 3 looks viable to me.

**Comment by rimrakhimov** [comment]
> We could make that integration with sig databases optional, and the base smart-contract-verifier service will still be "pure".

Yes, probably, in case of Option 2, we may implement that via some kind of middleware implementation, if we add Middleware trait as described in #132

**Comment by leviska** [comment]
So overview of current databases: 
1. https://sig.eth.samczsun.com/ We need only abi
![image](https://user-images.githubusercontent.com/14099557/194332995-07ba527b-2836-4a9a-8cc0-da03be83cd1c.png)
2. https://www.4byte.directory/docs/ We need only abi too
![image](https://user-images.githubusercontent.com/14099557/194333282-af50b59b-9aaa-4bfe-93de-7f3e4fcc48f4.png)

So I suggest we can make a minimalist api, which will accept contract address (for future proofing) and abi. Not sure what we might need in the future, so I think this will be enough for now
